modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java

```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
 
   }
 
+  /**
+   * Adds subtitles {@link PostEditingData.Subtitle} to the media package and sends the updated media package
+   * to the archive. If a subtitle flavor already exists, the subtitle is overwritten
+   *
+   * @param mediaPackage
+   *          the media package to at the SMIL catalog
+   * @param subtitles
+   *          the subtitles to be added
+   * @throws IOException
+   */
+  private void addSubtitleTrack(MediaPackage mediaPackage, List<PostEditingData.Subtitle> subtitles)
+          throws IOException, NullPointerException {
+    Objects.requireNonNull(subtitles, "Argument subtitles may not be null");
+
+    for (PostEditingData.Subtitle subtitle : subtitles) {
+      // Generate ID for new tracks
+      String subtitleId = UUID.randomUUID().toString();
+      String trackId = null;
+
+      // Check if subtitle already exists
+      for (Track t : mediaPackage.getTracks()) {
+        if (t.getFlavor().matches(subtitle.getFlavor())) {
+          logger.debug("Set Identifier for Subtitle-Track to: {}", t.getIdentifier());
+          subtitleId = t.getIdentifier();
+          trackId = t.getIdentifier();
+          break;
+        }
+      }
+
+      Track track = mediaPackage.getTrack(trackId);
+
+      // Memorize uri of the previous track file for deletion
+      URI oldTrackURI = null;
+      if (track != null) {
+        oldTrackURI = track.getURI();
+      }
+
+      // Put updated filename in working file repository and update the track.
+      InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8");
```

Comment:
It's a good idea to always close input streams

```suggestion
      try (InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8")) {
```